### DISCUSSION TITLE
Feature requests and ideas

### DISCUSSION THREAD

**Discussion by nvladimus** [mesoSPIM/mesoSPIM-control#54] [Ideas]
Hi, everyone,
We are actively adding new features in the `development` branch, see recent changes [here](https://github.com/mesoSPIM/mesoSPIM-control/blob/development/CHANGELOG.md). This is mostly based on our own usage of mesoSPIMs. If you have an idea how to significantly improve your productivity and user experience (UI, automation, file formats, error handling, etc), please post it here, or vote up already posted ideas. We will give them priority in the development!

**Comment by drchrisch** [comment]
We run a mesoSPIM V5 with PhysikInstumente (PI) stages equiped with stepper motors and individual controllers for each axes. In the README it is suggested to install *PiPython* version 1.3.5.37 to ensure proper connection to PI controllers. We noticed that a more recent version 2.3.0.3 available from pip works as well (at least with our setup!) and is easier to install using the __requirements_*.txt_ files and explicitly setting the version. However, the latest PiPython version 2.5.0.1 does not work and PI admitted that several files/commands are missing. They announced to have a working version again in 4-6 weeks (as of last week).

**Reply by nvladimus** [reply]
+1! Version `2.3.0.3` didn't work on our setup (BIMSB-Berlin), unfortunately. Please keep us posted about the feedback from PI, it would be great to have the latest `PiPython` working with mesoSPIM code.

**Reply by drchrisch** [reply]
A new pipython version is out (PIPython 2.5.1.3). I installed it from pip and it works with our setup (USB connection to controllers, multiple controllers). Check that it works also on other setups with 'quickcheck' ([https://github.com/drchrisch/mesoSPIM-extra/blob/main/quickcheck.py](url)).

@nvladimus @ffvoigt In case the updated pipython package also runs on your setups, you might consider simplifying mesoSPIM-control installation and modify the requirements file.

**Reply by nvladimus** [reply]
Hi, @drchrisch, I tested the Helmchen lab system with pip-installed PIPython 2.5.1.3, and after minor code tweaks, it works fine. I will update the installation instructions. Thank you!

**Comment by raacampbell** [comment]
A "Mark All" button in the acquisition manager has been requested by two of our users.

(BTW, I would like to be able to contribute more to the project. I am stretched right now, but I hope to be more active in the future).

**Reply by nvladimus** [reply]
Great to hear! 
Do you mean merging `Mark Current State` & `Mark ETL values`? Or XYZR positions?

**Reply by raacampbell** [reply]
I think the request was to merge all the Mark buttons. So one press will set the state of everything.

**Reply by nvladimus** [reply]
Mark X,Y,Z, Rotation, Focus, State (laser power, wl, filter), and ETL with one button? This may be too inclusive actually.

**Reply by raacampbell** [reply]
I find I do indeed click all the mark buttons for the first row when I start setting up a sequence.

**Reply by ffvoigt** [reply]
I think this is one of the cases where a reasonable simple default for new users plus a configurable option for power users might be necessary. Maybe a mark button plus a set of checkboxes for position, state etc that can be unchecked when necessary?

**Reply by nvladimus** [reply]
Done in [2f3314e](https://github.com/mesoSPIM/mesoSPIM-control/commit/2f3314e67a9b9c83371eb9a853fc102db4514dc4).

**Comment by raacampbell** [comment]
Integrate the "Live" and "Stop" buttons into one. So once "Live" is pressed it turns into "Stop".

**Comment by raacampbell** [comment]
Integrate laser control. We have two Toptica laser units and starting them up via the Toptica GUI is annoying. In the case of these lasers, I would envisage the MesoSPIM control software at startup just needs to connect to the two devices and set all lines to be under analog control. There could be four check boxes in the software to turn on the four lines. For simplicity we can just turn on the same line in both lasers, even if only one is going to be used.

**Reply by drchrisch** [reply]
Are you using TOPAS for laser control? I wonder whether there wasn't a way to set lasers to analog control at powering.

**Reply by raacampbell** [reply]
Yes. Maybe there is a way to do that but the lines still need enabling via the GUI.

**Reply by drchrisch** [reply]
Well, might be we use a different combination of laser and controller. Anyway, we managed to enable analog-only control.

**Reply by raacampbell** [reply]
But you still use the GUI to enable the laser lines, no?

**Reply by drchrisch** [reply]
No, not at all.
We configured the laser using TOPAS (direct USB connection from computer) to react to analog input. If I remember correctly, the laser power has to be set to 100% (I'm not sure, can't access the instrument right now). Do not forget to save settings. Then it is just to switch on laser before starting mesoSPIM-control.

**Reply by raacampbell** [reply]
OK, I'll check it out. At the moment it disables all the lines each time the key is turned off.

**Reply by ffvoigt** [reply]
As far as I know, the USZ and ZMB Toptica MLEs also keep their state when power cycling. Did you wire up the digital enable inputs as well? I'm a little wary about trying to start integrating laser-specific code as there are so many different manufacturers out there and all have different SDKs. The reason for using analog outputs for laser control was to keep the laser interface simple - even though it requires more AO channels.

**Reply by raacampbell** [reply]
Good to know. I'll sort it out as you describe, I don't recall whether I wired up the digital lines.

**Reply by nvladimus** [reply]
I agree with Fabian, keeping general AO/DO waveforms for laser control allows more versatility in lasers to choose from. We have Cobolt F-Flex, we start it in native software, turn lasers we need ON, analog control ON, and it works just fine. No worries about broken code etc.

**Reply by raacampbell** [reply]
Yes, I agree.

**Reply by raacampbell** [reply]
To confirm: all I had to do was hit "Save Settings" and the lasers start in the desired state. The digital lines aren't needed for the Toptica laser.

**Comment by raacampbell** [comment]
Start control software via an Icon on the desktop.

**Reply by nvladimus** [reply]
Solved.

**Comment by raacampbell** [comment]
In samples that are not very well cleared, we set up the ETL as follows:
- Reduce galvo scan angle until at least on of the turn-arounds is visible. 
- Reduce laser power as much as is reasonable
- Set ETL amplitude to zero and tweak offset
- Enable ETL sweep and tweak ETL amplitude until turn-around looks equally fat throughout. 
- Return galvo amplitude to a more reasonable value

The above works pretty well but is harder to do than it needs to be because the controls for the galvo amplitude are on a different tab. Ideally I would like to have a "show turnaround" checkbox that automatically reduces galvo amplitude to a suitable value. Presumably we just need to obtain such a value for zoom 1 and save that in the settings. Then we scale it according to zoom. We could even scale the power automatically with zoom to reduce bleaching. 

It would also be _really_ helpful if there was a laser power edit box in the main GUI. This would be faster and more direct than a slider. e.g. User types 3 into the box and laser is set to 3%. Dragging the slider would update the value in the box.

**Reply by nvladimus** [reply]
Laser power edit box in the main GUI: Solved.
Galvo amp scaling with Zoom: solved.

**Comment by raacampbell** [comment]
List FOV and pixel size in the preview image window title bar.

**Reply by drchrisch** [reply]
Good point.

**Reply by nvladimus** [reply]
BTW, you can measure FOV or an object size using resizable Box overlay, in the lower left corner of the image window. Available in the `development` branch.

**Comment by raacampbell** [comment]
I think for co-alignment of the sheets, a red/green overlay would be easier to use than the switching display.

**Comment by nvladimus** [comment]
ETL parameters automatically adjusted with one click of a button.

**Comment by nvladimus** [comment]
Autofocus

**Reply by nvladimus** [reply]
Solved

**Comment by nvladimus** [comment]
Galvo amplitude (light-sheet height) auto-adjusted to match camera FOV, at various zoom factors.

**Reply by drchrisch** [reply]
Very good idea!

**Reply by raacampbell** [reply]
I think if you do that then you also need to scale the laser power to match zoom. The galvo adjustment should be optional in case there are problems with bleaching at the turn-arounds.

**Reply by nvladimus** [reply]
Solved.

**Comment by JinlongL** [comment]
Hi. We are trying to adjust the ETL_Amplitude to above 2.0 V. Do you know where and how in the code we may change the maximum ETL amplitude?  Thank you so much.

**Reply by nvladimus** [reply]
Hi, you need to open the file `mesoSPIM_MainWindow.ui` in text editor or **Qt Designe**r program, and change `LeftETLAmplitudeSpinBox` maximum value. If you are editing the text file directly, look for
```
<property name="maximum">
<double>2.000000000000000</double>
</property>
```
or in Qt Designer:
![image](https://user-images.githubusercontent.com/10835134/148804311-700bbf46-fbfd-4ca9-9dca-72a42b21458b.png)

It is better open an issue for this kind of questions. This branch is more for free-form discussions.

**Reply by JinlongL** [reply]
Thank you so so much for the help. And I am deeply sorry about asking the question in the wrong branch. Thank you again.    :).

**Reply by nvladimus** [reply]
🙂
